### PR TITLE
Support UseSgprForGRO for dtva

### PR DESCRIPTION
**Pull request by mengzcai**
Done: support for DTVA/B+UseSgprForGRO

### CODE REVIEW

**Review by aazz44ss — DISMISSED** [review]

File: tensilelite/Tensile/Tests/common/gemm/use_sgpr_for_gro_with_dtva.yaml [hunk 1]
```diff
@@ -0,0 +1,1130 @@
+TestParameters:
+  marks: [skip-gfx900, skip-gfx906, skip-gfx908, skip-gfx90a, skip-gfx1010, skip-gfx1011, skip-gfx1012, skip-gfx1030, skip-gfx1100, skip-gfx1101, skip-gfx1102, skip-gfx1200, skip-gfx1201] # not supported by arch
+
+GlobalParameters:
```
Comment (jichangjichang) [inline tensilelite/Tensile/Tests/common/gemm/use_sgpr_for_gro_with_dtva.yaml]: don't create another testing yaml.
You should directly add the test into original dtv testing yaml

Comment (mengzcai) [reply tensilelite/Tensile/Tests/common/gemm/use_sgpr_for_gro_with_dtva.yaml]: fixed

File: tensilelite/Tensile/SolutionStructs.py [hunk 2]
```diff
@@ -3960,20 +3960,26 @@ def calcEpilogueTurns(factorDims: List) -> int:
...
+      if state["_UseSgprForGRO"] == -1:
+        if numLoadsA + numLoadsB + numLoadsM > 35 or state["ProblemType"]["SwizzleTensorA"] or state["ProblemType"]["SwizzleTensorB"]:
+          #print "info: Disabling UseSgprForGRO since predicting too many SGPR will be used"
+          state["_UseSgprForGRO"] = 0
+        else:
+          state["_UseSgprForGRO"] = 1
 
+      if state["_UseSgprForGRO"] == 1:
+        if numLoadsA + numLoadsB + numLoadsM > 35:
+          reject(state, "Too many SGPR will be used for UseSgprForGRO")
```
Comment (aazz44ss) [inline tensilelite/Tensile/SolutionStructs.py]: We don't need to reject it in SolutionStruct. If it uses too many sgpr, it should be rejected when checking vgpr usage in final stage.

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 3]
```diff
@@ -2549,8 +2549,33 @@ def graFinalOffsets(self, kernel, tP):
...
+                module.add(singleModule)
+      else:
+        numOfVW = kernel["MIWaveTile%c"%tc]//kernel["VectorWidth%s"%tc] if kernel["VectorWidth%c"%tc]>1 else 1
+        sizeOfaPart = tP["nrp"]//numOfVW
+        # swap para and perp
+        for para in range(0, tP["nrc"]):
+          for sPara in range(0, int(tP["nrcv"]/tP["nrcvpi"])):
+            for sPerp in range(0, tP["nrpv"]*numOfVW):
+              for perp in range(0, sizeOfaPart): #tP["nrp"]==tP["nrt"]
+            #for perp in range(0, tP["nrp"]): #tP["nrp"]==tP["nrt"]
```
Comment (jichangjichang) [inline tensilelite/Tensile/KernelWriterAssembly.py]: clean unused codes

File: tensilelite/Tensile/SolutionStructs.py (hunk 2 above)
Comment (mengzcai) [reply tensilelite/Tensile/SolutionStructs.py]: fixed, thx!

File: tensilelite/Tensile/KernelWriterAssembly.py (hunk 3 above)
Comment (mengzcai) [reply tensilelite/Tensile/KernelWriterAssembly.py]: fixed, thx!

**Review by jichangjichang — APPROVED** [review]

**Review by hcman2 — APPROVED** [review]

**Pull request merged**